generate_route.py

```diff
@@ -81,36 +81,112 @@ def __init__(self, start_location, end_location):
     self.polyline = None
 
   def create_route(self):
+    """Create a route by requesting from Maps API and then adding altitudes/xyz to each point.
+
+    sets attributes for the class based on API response and then calls add_altitudes() to 
+    request elevation data for each point, and then add xyz conversion to each point
+    """
     locations,self.distances,self.durations,self.polyline = request_directions(self.start_location.get_lat_lon_tuple(),self.end_location.get_lat_lon_tuple())
     route = []
     for location in locations:
       latitude = location[0]
       longitude = location[1]
       route.append(Location(latitude,longitude))
     self.route = route
+    self.add_altitudes()
   
   def add_altitudes(self):
+    """Add altitudes to each point in self.route using Maps Elevation API.
+
+    sets elevation data for each point, and then adds xyz conversion from lla to each point
+    """
     locations = []
     for location in self.route:
       locations.append(location.get_lat_lon_tuple())
     elevations = request_elevations(locations)
     for location,elevation in zip(self.route,elevations):
       location.altitude = elevation
+      location.add_XYZ()
+
+  def write_route(self,file_name):
+    """write route into csv with each line as x,y,z.
+
+    """
+    write_array = []
+    for location in self.route:
+      write_array.append(location.get_xyz_tuple())
+    write_to_csv(file_name,write_array)
+
+class TimedRoute(Route):
+  """An object for a route that has a desired speed and point frequency
+
+  Attributes:
+      start_location: a Location object for the start of the route
+      end_location: a Location object for the end of the route
+      speed: how fast the person moves through the route
+      frequency: how many points per second the timed route should have (Hz)
+      route: a list of Location objects for each point on the route
+      distances: a list of distances between each pair of consecutive locations
+      durations: a list of durations between each pair of consecutive locations
+      polyline: an encoded format for the route given by the Maps API for ease of drawing
+  """
+  def __init__(self, start_location, end_location,speed,frequency):
+    Route.__init__(self,start_location,end_location)
+    self.speed = speed
+    self.frequency = frequency
+
+  def create_route(self):
+    Route.create_route(self)
+    self.upsample_route()
   
   def upsample_route(self):
-    pass
+    """Upsample the TimedRoute such that the number of points aligns with the desired
+    speed and frequency.
+
+    for each consecutive set of points, the change in lat,lon,alt is divided by
+    split amongst the number of new points that need to be created so that there
+    is roughly an equal distance (1/points_per_meter) between each of the points in the 
+    upsampled route.
+    """
+    points_per_meter = self.frequency/self.speed
+    new_route = []
+
+    # fill first 10 cycles with first location
```

Comment:
Chris had this in his prototype and I assumed it was for some reason that had to do with having a consistent location at the beginning to lock on to or something like that, but I feel like it can probably just be taken out

Reply:
Chris says that it might be because it's easier for the device to lock on if the device "stands still" for a while. Let's leave it in for now and add a TODO to check if this is actually necessary.

Reply:
Chris said that this might be to make it easier for the device to lock on because it's not "in motion".